simulator/main.rs

```diff
@@ -67,6 +123,433 @@ enum Value {
+            // An AND for true requires each of its children to be true
+            // An AND for false requires at least one of its children to be false
+            if *predicate_value {
+                Predicate::And(
+                    (0..rng.gen_range(1..=3))
+                        .map(|_| SimplePredicate::arbitrary_from(rng, &(*table, true)).0)
+                        .collect(),
+                )
+            } else {
+                // Create a vector of random booleans
+                let mut booleans = (0..rng.gen_range(1..=3))
+                    .map(|_| rng.gen_bool(0.5))
+                    .collect::<Vec<_>>();
+
+                let len = booleans.len();
+
+                // Make sure at least one of them is false
+                if booleans.iter().all(|b| *b) {
+                    booleans[rng.gen_range(0..len)] = false;
+                }
+
+                Predicate::And(
+                    booleans
+                        .iter()
+                        .map(|b| SimplePredicate::arbitrary_from(rng, &(*table, *b)).0)
+                        .collect(),
+                )
+            }
+        } else {
+            // An OR for true requires at least one of its children to be true
+            // An OR for false requires each of its children to be false
+            if *predicate_value {
+                // Create a vector of random booleans
+                let mut booleans = (0..rng.gen_range(1..=3))
+                    .map(|_| rng.gen_bool(0.5))
+                    .collect::<Vec<_>>();
+                let len = booleans.len();
+                // Make sure at least one of them is true
+                if booleans.iter().all(|b| !*b) {
+                    booleans[rng.gen_range(0..len)] = true;
+                }
+
+                Predicate::Or(
+                    booleans
+                        .iter()
+                        .map(|b| SimplePredicate::arbitrary_from(rng, &(*table, *b)).0)
+                        .collect(),
+                )
+            } else {
+                Predicate::Or(
+                    (0..rng.gen_range(1..=3))
+                        .map(|_| SimplePredicate::arbitrary_from(rng, &(*table, false)).0)
+                        .collect(),
+                )
+            }
+        })
+    }
+}
+
+impl ArbitraryFrom<Table> for Predicate {
+    fn arbitrary_from<R: Rng>(rng: &mut R, table: &Table) -> Self {
+        let b = rng.gen_bool(0.5);
+        CompoundPredicate::arbitrary_from(rng, &(table, b)).0
+    }
+}
+
+impl ArbitraryFrom<(&str, &Value)> for Predicate {
+    fn arbitrary_from<R: Rng>(rng: &mut R, (column_name, value): &(&str, &Value)) -> Self {
+        match rng.gen_range(0..3) {
+            0 => Predicate::Eq(column_name.to_string(), (*value).clone()),
+            1 => Predicate::Gt(column_name.to_string(), LTValue::arbitrary_from(rng, *value).0),
+            2 => Predicate::Lt(column_name.to_string(), LTValue::arbitrary_from(rng, *value).0),
```

Comment:
We need to also test where `WHERE` does not exist too right?

Reply:
That makes sense, I think we ideally would like this type to reach the whole of `Select` type in the limbo AST. 

```rust
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Select {
    /// CTE
    pub with: Option<With>,
    /// body
    pub body: SelectBody,
    /// `ORDER BY`
    pub order_by: Option<Vec<SortedColumn>>, // ORDER BY term does not match any column in the result set
    /// `LIMIT`
    pub limit: Option<Limit>,
}

/// `SELECT` body
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SelectBody {
    /// first select
    pub select: OneSelect,
    /// compounds
    pub compounds: Option<Vec<CompoundSelect>>,
}
...
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CompoundSelect {
    /// operator
    pub operator: CompoundOperator,
    /// select
    pub select: OneSelect,
}

/// Compound operators
// https://sqlite.org/syntax/compound-operator.html
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum CompoundOperator {
    /// `UNION`
    Union,
    /// `UNION ALL`
    UnionAll,
    /// `EXCEPT`
    Except,
    /// `INTERSECT`
    Intersect,
}

/// `SELECT` core
// https://sqlite.org/syntax/select-core.html
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum OneSelect {
    /// `SELECT`
    Select {
        /// `DISTINCT`
        distinctness: Option<Distinctness>,
        /// columns
        columns: Vec<ResultColumn>,
        /// `FROM` clause
        from: Option<FromClause>,
        /// `WHERE` clause
        where_clause: Option<Expr>,
        /// `GROUP BY`
        group_by: Option<GroupBy>,
        /// `WINDOW` definition
        window_clause: Option<Vec<WindowDef>>,
    },
    /// `VALUES`
    Values(Vec<Vec<Expr>>),
}
```